src/lib/utils/ct_utils.h

```diff
@@ -107,6 +107,54 @@ constexpr inline T value_barrier(T x) {
    }
 }
 
+/**
+* A Choice is used for constant-time conditionals.
+*
+* Internally it always is either all 0 bits or all 1 bits.
+*/
+class Choice final {
+   public:
+      /**
+      * If v == 0 return an unset (false) Choice, otherwise a set Choice
+      */
+      template <typename T>
+      constexpr static Choice from_int(T v)
+         requires(std::is_unsigned<T>::value)
+      {
+         const auto v_is_0 = ct_is_zero<T>(value_barrier<T>(v));
+         return Choice(ct_is_zero<uint32_t>(static_cast<uint32_t>(v_is_0)));
+      }
+
+      template <typename T>
+      constexpr T choose(T a, T b)
+         requires(std::is_unsigned<T>::value)
+      {}
+
+      constexpr Choice operator!() const { return Choice(~value()); }
+
+      constexpr Choice operator&&(const Choice& other) const { return Choice(value() & other.value()); }
+
+      constexpr Choice operator||(const Choice& other) const { return Choice(value() | other.value()); }
+
+      constexpr Choice operator!=(const Choice& other) { return Choice(value() ^ other.value()); }
+
+      /// Unsafe conversion to bool
+      constexpr bool as_bool() const { return m_value != 0; }
```

Comment:
In what sense is this unsafe? Given that the compiler shouldn't be able to reason about `m_value` it shouldn't be able to reason about the `bool` either. Isn't it only the potential downstream usage of the bool that makes it "unsafe"?

Reply:
Yes and no :)

`bool` has a constrained value (it is either `true` or `false`) which provides a choke point on the possible range of values, so it's entirely possible for the compiler to reason about the cases. Consider a hypothetical constant time call:

```
bool b = choice.as_bool();
uint32_t f = expensive_and_pure_fn(inputs);
uint32_t mask = b * 0xFFFFFFFF;
return f & mask;
```

The compiler doesn't know what `b` is, but it knows it is either `true` or `false`. From there it is easy to see that `f` can be omitted entirely if `b` is false, so transforms into

```
return (b) ? expensive_and_pure_fn(inputs) : 0;
```

That's why the internals of this are using a uint32_t - be a large enough state that, assuming the compiler is willing to forget what it knows about the contents, it won't try to optimize on a case-by-case enumeration of the possible states.

Reply:
I added a comment in the source explaining more about why the conversion itself is potentially unsafe

Reply:
So the relevant point is that, even if the compiler cannot infer the actual `bool` value, it might still optimize for the case that it is one or the other. Docstring is also making this clear. Thanks.